### PR TITLE
Issue126

### PR DESCRIPTION
**Pull request by jkirsh2**
- added cumulative energy delta tracking to vehicles and stations
- include energy dispensed/expended in summary report

### CODE REVIEW

File: nrel/hive/model/station/station.py [hunk 1]
```diff
@@ -408,6 +413,19 @@ def receive_payment(self, currency_received: Currency) -> Station:
...
         return replace(self, balance=self.balance + currency_received)
 
+    def tick_energy_dispensed(self, delta_energy: immutables.Map[EnergyType, float]) -> Station:
+        """
+        adds energy dispensed to vehicle
+
+        :param delta_energy: the energy dispensed for a charge event
+        :return: the updated Station
+        """
+        energy_dispensed = {
```
Comment (robfitzgerald) [inline nrel/hive/model/station/station.py:423]: wrap with immutables.Map

File: nrel/hive/model/vehicle/vehicle.py [hunk 2]
```diff
@@ -207,6 +213,26 @@ def tick_distance_traveled_km(self, delta_d_km: Kilometers) -> Vehicle:
...
         return replace(self, distance_traveled_km=self.distance_traveled_km + delta_d_km)
 
+    def tick_energy_expended(self, delta_energy: immutables.Map[EnergyType, float]) -> Vehicle:
+        """
+        adds energy expenditure to vehicle
+
+        :param delta_energy:
+        :return:
+        """
+        energy_expended = {k: self.energy_expended[k] + delta_energy[k] for k in self.energy.keys()}
```
Comment (robfitzgerald) [inline nrel/hive/model/vehicle/vehicle.py:223]: wrap with immutables.Map

File: nrel/hive/model/vehicle/vehicle.py [hunk 3]
```diff
@@ -207,6 +213,26 @@ def tick_distance_traveled_km(self, delta_d_km: Kilometers) -> Vehicle:
...
+        return replace(self, energy_expended=energy_expended)
+
+    def tick_energy_gained(self, delta_energy: immutables.Map[EnergyType, float]) -> Vehicle:
+        """
+        adds energy gain to vehicle
+
+        :param delta_energy:
+        :return:
+        """
+        energy_gained = {k: self.energy_gained[k] + delta_energy[k] for k in self.energy.keys()}
```
Comment (robfitzgerald) [inline nrel/hive/model/vehicle/vehicle.py:233]: ditto

**Review by robfitzgerald — CHANGES_REQUESTED** [review]
looks great. so satisfying to see this now in the logs:

```
│ Total kWh Expended By Vehicles               │ 676.09 kWh │
│ Total Gasoline Expended By Vehicles          │ 70.24 Gal  │
│ Total kWh Dispensed By Stations              │ 872.03 kWh │
│ Total Gasoline Dispensed By Stations         │ 7.0 Gal    │
```

a few minor requests to wrap {}'s in Map(). thank you!

**Comment by jhoshiko** [comment]
I wrapped the dictionaries since the two volunteers working on this had to leave, but we'll need to reformat `vehicle.py` and `station.py` with black since I can't do that through github's file editor. Talking with the team, it might make sense to add running black automatically with a pre-commit hook.

**Comment by robfitzgerald** [comment]
i'll take a stab at fixing this, likely pushing up a new PR with the changes + formatting

**Comment by nreinicke** [comment]
Closing as this was included in #179

**Pull request closed without merging**